src/vs/workbench/contrib/inlineChat/browser/inlineChatController.ts

```diff
@@ -234,6 +244,54 @@ export class InlineChatController implements IEditorContribution {
 		return INLINE_CHAT_ID;
 	}
 
+	private _installProviderEnablementListeners() {
+		this._store.add(this._inlineChatService.onDidChangeProviders(async (e) => {
+			if (e.added) {
+				this._providerEnablementEvents.set(e.added, e.added.onDidChangeEnablementStatus(async () => {
+					if ((await this._enabledProvider) === e.added) {
```

Comment:
this is likely flawed when receiving multiple events while still resolving `this._enabledProvider`. As a role of thumb: don't await during event handling or make sure assumptions before awaiting still hold after

Reply:
I wonder what to do here then. I guess you just assume because it's still resolving that whatever is resolving is an older event and you want this to supersede it anyways? The worry is that you have two providers one that is disabled and one that is enabled and you always want to keep the enabled one around regardless if this toggle their enablements or anything like that .